src/passes/staticArrayIndexer.ts

```diff
@@ -0,0 +1,207 @@
+    if (!hasIndexAccess(node, ast)) {
+      return this.visitExpression(node, ast);
+    }
+
+    const identifier = getReferenceDeclaration(node, ast);
+
+    assert(identifier.vReferencedDeclaration instanceof VariableDeclaration);
+    const parentFunction = node.getClosestParentByType(FunctionDefinition);
+    assert(
+      parentFunction !== undefined,
+      'Calldata types must be enclosed in a function definition',
+    );
+
+    const refId = identifier.referencedDeclaration;
+
+    let refVarDecl = this.staticArrayAccesed.get(refId);
+    if (refVarDecl === undefined) {
+      refVarDecl = this.initMemoryArray(identifier, parentFunction, ast);
+      this.staticArrayAccesed.set(refId, refVarDecl);
+    }
+
+    const indexReplacement = this.setExpressionToMemory(node, refVarDecl, ast);
+    ast.context.register(node);
+    ast.replaceNode(node, indexReplacement);
+  }
+
+  private setExpressionToMemory(
+    expr: Expression,
+    refVarDecl: VariableDeclaration,
+    ast: AST,
+  ): Identifier | IndexAccess | MemberAccess {
+    const exprType = getNodeType(expr, ast.compilerVersion);
+    const exprMemoryType = specializeType(generalizeType(exprType)[0], DataLocation.Memory);
+
+    let replacement: Identifier | IndexAccess | MemberAccess;
+    if (expr instanceof Identifier) {
+      replacement = createIdentifier(refVarDecl, ast, DataLocation.Memory);
+    } else if (expr instanceof IndexAccess) {
+      assert(expr.vIndexExpression instanceof Expression);
+      replacement = new IndexAccess(
+        expr.id,
+        expr.src,
+        generateExpressionTypeString(exprMemoryType),
+        this.setExpressionToMemory(expr.vBaseExpression, refVarDecl, ast),
+        cloneASTNode(expr.vIndexExpression, ast),
+      );
+    } else if (expr instanceof MemberAccess) {
+      replacement = new MemberAccess(
+        expr.id,
+        '',
+        generateExpressionTypeString(exprMemoryType),
+        this.setExpressionToMemory(expr.vExpression, refVarDecl, ast),
+        expr.memberName,
+        expr.referencedDeclaration,
+      );
+    } else {
+      throw new NotSupportedYetError(
+        `Static array index access with nested expression ${printNode(expr)} is not supported yet`,
+      );
+    }
+    ast.context.unregister(expr);
+    return replacement;
+  }
+
+  private initMemoryArray(
+    identifier: Identifier,
+    parentFunction: FunctionDefinition,
+    ast: AST,
+  ): VariableDeclaration {
+    const refId = identifier.referencedDeclaration;
+    const memoryType = specializeType(
+      generalizeType(getNodeType(identifier, ast.compilerVersion))[0],
+      DataLocation.Memory,
+    );
+
+    const varDecl = new VariableDeclaration(
+      ast.reserveId(),
+      '',
+      false,
+      false,
+      `memory_${identifier.name}`,
+      parentFunction?.id,
+      false,
+      DataLocation.Memory,
+      StateVariableVisibility.Internal,
+      Mutability.Mutable,
+      generateExpressionTypeString(memoryType),
```

Comment:
VariableDeclarations have different typestrings to expressions as they don't contain the memory location. I think what you'll want to do here is to both not specialise `memoryType` and just call `.pp()` on it rather than `generateExpressionTypeString`